### ISSUE TITLE
computational time issue of Java generics

### ISSUE THREAD

**Issue by wanrong3** [doxygen/doxygen#6382]
hello, I have the very simple class with generics, detailed as below. When running the step of computing class usage relations, it is never run to the end. 

```
package a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z;
import java.util.*;

public class A0 {

List<com.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.A0> b = new ArrayList<>();

String c = null;

}
```

 After logging something, i found it ran function resolveTypeDef in util.cpp file recursively, and never ran out. So it can not normally finish when it is at the situation.

**Comment by albert-github** [comment]
There are a number of questions to be answered here:
- is it really an endless loop
  - in `doxygen.cpp` the routine `resolveTypeDef` is called in a `for` loop, so this shouldn't result in an endless loop
  - in `util.cpp`  in the routine `resolveTypeDef` there is in a `while` loop an other call to `resolveTypeDef`, so potential endless loop.
  - some timings, instead of using all levels I used a subset:
     - a.q.r.s.t.u.v.w.x.y.z: 5.7 sec
     - a.p.q.r.s.t.u.v.w.x.y.z: 22.9 sec
     - a.o.p.q.r.s.t.u.v.w.x.y.z: 93.5 sec
     - a.n.o.p.q.r.s.t.u.v.w.x.y.z: 374.4 sec

     so roughly with each extra level a factor of 4.
- what is the situation realistic (the example is in my opinion exaggerating, but this is OK as it points out the problem more clearly). Here we have a "simple" case but what are the effects on more complex / larger cases.

For this it is certainly worthwhile to investigate.

**Comment by wanrong3** [comment]
Hi, Albert, Thanks for your reply. we also found that removing `List<` and ` >` the routine returns immediately. Is correct for doxygen taking the arguments of java generics as the normal class? And if they are been taken as normal class, the routine might returns fast, I guess.  It is not clear that why the mechanism between processing normal class and arguments of generics is so different.

**Comment by WilliamsPhil** [comment]
Hi, Albert,

Thank you for your kind reply. The problem raised by wanrong is considerable and worth to be handled.

1.	The routine resolveTypeDef can run to end but cost too many loops and too much time. As your example, parsing a java class costs 374.4 sec which is time-consuming and unacceptable.

2.	Before taking actions, one thing should be clarified, in my opinion, that why List< com.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.A0 > is modified to be List< com.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a::b::c::d::e::f::g::h::i::j::k::l::m::n::o::p::q::r::s::t::u::v::w::x::y::z::A0 > for resolveTypeDef? What is the aim?

3.	As mentioned above, why processing mechanism for normal classes and arguments of generics is different?

**Comment by Christian390** [comment]
Hi, I have the same problem with some of my classes. I don't know if it helps you, but the first example does not compile: "com." should be removed from "List<com.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.A0> b = new ArrayList<>();"

And I have one more minimal example that does not come to end with computation of class usage relations. I let it run 30 minutes. EXTRACT_STATIC has to be enabled to reproduce the problem.

    package aaa.aaa.aaa.aaa.aaa.aaa.aaa.aaa.aaa;

    import java.util.ArrayList;
    import java.util.List;

    public class Test
    {
       protected Test.InnerClass myMember;

       public Test.InnerClass getMyMember()
       {
          return myMember;
       }

       public void setMyMember(Test.InnerClass value)
       {
          this.myMember = value;
       }

       public static class InnerClass
       {
          protected List<Test.InnerClass.Group> group;

          public List<Test.InnerClass.Group> getGroup()
          {
             if (group == null)
             {
                group = new ArrayList<Test.InnerClass.Group>();
             }
             return this.group;
          }

          public static class Group
          {
          }
       }
    }

**Comment by albert-github** [comment]
@Christian390 
- I don't know by head whether or not this has been resolved yet but which version of doxygen did you use?

**Comment by Christian390** [comment]
I'm using 1.9.3 on Windows.

**Comment by Christian390** [comment]
I could reproduce the problem with both examples with Doxygen 1.9.3 on Windows.

**Comment by doxygen** [comment]
@Christian390 Can you please verify if the problem is fixed with the referenced commit?
You could try one of the Windows binaries built by [GitHub actions](https://github.com/doxygen/doxygen/actions/runs/1733357594) if you don't want to build doxygen yourself.

**Comment by albert-github** [comment]
@doxygen 
I tried the (insane) example in the original question in Java and this runs fast now.
I tried an analogous (insane) example for Cpp
```
///the namespace
namespace a::b::c::d::e::f::g::h::i::j::k::l::m::n::o::p::q::r::s::t::u::v::w::x::y::z
{
/// the class
class A0
{

/// the var
List<a::b::c::d::e::f::g::h::i::j::k::l::m::n::o::p::q::r::s::t::u::v::w::x::y::z::A0> b = new ArrayList<>();
}
}
```
but here we still see a speed problem (the large example chokes). I tried with a reduced set, starting wit test `q::...::z` and prepending each time a letter and at that moment we see the time increases roughly with each extra level by a factor of 4. With a sensible number of levels this will not really give problems.

**Comment by doxygen** [comment]
@albert-github Indeed for C++ nothing has improved yet. I'll have a look how to optimize the routine. The duration should not grow exponentially as it seems to do now.

**Comment by doxygen** [comment]
@albert-github I've now also excluded templates from typedef processing, as in C++ a typedef cannot be a template.

**Comment by albert-github** [comment]
@doxygen thanks, it runs now very fast.

**Comment by Christian390** [comment]
I'm happy to see that you found a solution for that problem. Thanks a lot!
I tested https://github.com/doxygen/doxygen/actions/runs/1736709043 and it runs very fast with the example as well as with my real code.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
